File: download-images.py
#!/usr/bin/env python3
"""
Script to download article images from URLs

Usage: python3 download-images.py [ARTICLE_URL] [ARTICLE_SLUG]

Example:
python3 download-images.py https://bizee.com/articles/can-a-minor-own-a-business can-a-minor-own-a-business
"""

import sys
import os
import re
import json
import urllib.request
import urllib.parse
from pathlib import Path
from datetime import datetime

class ImageDownloader:
    def __init__(self, article_url, article_slug):
        self.article_url = article_url
        self.article_slug = article_slug
        # Get project base directory (two levels up from script)
        script_dir = Path(__file__).parent
        self.base_dir = script_dir.parent / 'public' / 'assets'
        self.featured_path = self.base_dir / 'articles' / 'featured'
        self.main_content_path = self.base_dir / 'articles' / 'main-content'

        # Create directories
        self.featured_path.mkdir(parents=True, exist_ok=True)
        self.main_content_path.mkdir(parents=True, exist_ok=True)

    def download_image(self, image_url, destination_path):
        """Downloads an image from a URL"""
        try:
            # Create request with headers to avoid blocking
            req = urllib.request.Request(
                image_url,
                headers={
                    'User-Agent': 'Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36'
                }
            )

            with urllib.request.urlopen(req) as response:
                image_data = response.read()

                # Save image
                with open(destination_path, 'wb') as f:
                    f.write(image_data)

                return True
        except Exception as e:
            print(f"✗ Error downloading {image_url}: {e}")
            return False

    def extract_images_from_html(self, html):
        """Extracts image URLs from HTML"""
        images = []

        # Find images in <img> tags
        img_pattern = r'<img[^>]+src=["\']([^"\']+)["\'][^>]*>'
        matches = re.findall(img_pattern, html, re.IGNORECASE)
        images.extend(matches)

        # Find images in data-src (lazy loading)
        data_src_pattern = r'data-src=["\']([^"\']+)["\']'
        data_matches = re.findall(data_src_pattern, html, re.IGNORECASE)
        images.extend(data_matches)

        # Find images in background-image attributes
        bg_pattern = r'background-image:\s*url\(["\']?([^"\']+)["\']?\)'
        bg_matches = re.findall(bg_pattern, html, re.IGNORECASE)
        images.extend(bg_matches)

        # Remove duplicates and filter
        unique_images = []
        seen = set()

        for img in images:
            # Filter small images (icons, etc.)
            if any(skip in img.lower() for skip in ['icon', 'logo', 'avatar', 'favicon']):
                continue

            # Normalize URL
            img = self.normalize_url(img)

            if img and img not in seen:
                seen.add(img)
                unique_images.append(img)

        return unique_images

    def normalize_url(self, url):
        """Normalizes a relative URL to absolute"""
        if not url:
            return None

        # If already absolute, return as is
        if url.startswith('http://') or url.startswith('https://'):
            return url

        # If relative, convert to absolute
        parsed = urllib.parse.urlparse(self.article_url)
        base = f"{parsed.scheme}://{parsed.netloc}"

        if url.startswith('/'):
            return base + url
        else:
            return base + '/' + url

    def get_image_extension(self, url):
        """Gets image extension from URL"""
        parsed = urllib.parse.urlparse(url)
        path = parsed.path
        ext = os.path.splitext(path)[1].lower()

        if not ext:
            # Try to detect from content-type or use png as default
            ext = '.png'

        # Normalize extensions
        if ext == '.jpg':
            ext = '.jpeg'

        return ext

    def generate_filename(self, image_url, index, is_featured=False):
        """Generates a filename for the image"""
        ext = self.get_image_extension(image_url)

        if is_featured:
            filename = f"{self.article_slug}{ext}"
            return self.featured_path / filename
        else:
            filename = f"{self.article_slug}-{index}{ext}"
            return self.main_content_path / filename

    def fetch_article_html(self):
        """Gets article HTML"""
        try:
            req = urllib.request.Request(
                self.article_url,
                headers={
                    'User-Agent': 'Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36'
                }
            )

            with urllib.request.urlopen(req) as response:
                return response.read().decode('utf-8')
        except Exception as e:
            print(f"✗ Error getting HTML: {e}")
            return None

    def process_article(self):
        """Processes article and downloads all images"""
        print(f"=== Downloading images for: {self.article_slug} ===\n")
        print(f"Article URL: {self.article_url}\n")

        html = self.fetch_article_html()
        if not html:
            return

        images = self.extract_images_from_html(html)

        if not images:
            print("No images found in article.")
            return

        print(f"Found {len(images)} image(s):\n")

        downloaded_images = []

        for i, image_url in enumerate(images):
            print(f"Processing ({i+1}/{len(images)}): {image_url}")

            is_featured = (i == 0)
            destination_path = self.generate_filename(image_url, i, is_featured)

            if self.download_image(image_url, destination_path):
                # Save relative path from public/assets
                relative_path = str(destination_path.relative_to(self.base_dir))

                downloaded_images.append({
                    'original_url': image_url,
                    'local_path': relative_path.replace('\\', '/'),
                    'type': 'featured' if is_featured else 'main-content',
                    'index': i
                })
                print(f"✓ Descargada: {relative_path}\n")
            else:
                print()

        # Generate report
        self.generate_report(downloaded_images)

    def generate_report(self, downloaded_images):
        """Generates a JSON report of downloaded images"""
        report = {
            'article_slug': self.article_slug,
            'article_url': self.article_url,
            'downloaded_at': datetime.now().isoformat(),
            'images': downloaded_images
        }

        script_dir = Path(__file__).parent
        report_path = script_dir / f'image-mapping-{self.article_slug}.json'

        with open(report_path, 'w', encoding='utf-8') as f:
            json.dump(report, f, indent=2, ensure_ascii=False)

        print(f"✓ Reporte guardado en: {report_path}")

if __name__ == '__main__':
    if len(sys.argv) < 3:
        print("Usage: python3 download-images.py [ARTICLE_URL] [ARTICLE_SLUG]")
        print("\nExample:")
        print("python3 download-images.py https://bizee.com/articles/can-a-minor-own-a-business can-a-minor-own-a-business")
        sys.exit(1)

    article_url = sys.argv[1]
    article_slug = sys.argv[2]

    downloader = ImageDownloader(article_url, article_slug)
    downloader.process_article()

File: download-images.php
<?php

/**
 * Script to download article images from URLs
 *
 * Usage: php download-images.php [ARTICLE_URL] [ARTICLE_SLUG]
 * Example: php download-images.php https://bizee.com/articles/can-a-minor-own-a-business can-a-minor-own-a-business
 */

class ImageDownloader
{
    private $articleUrl;
    private $articleSlug;
    private $assetsBasePath;
    private $featuredPath = 'articles/featured';
    private $mainContentPath = 'articles/main-content';

    public function __construct($articleUrl, $articleSlug)
    {
        $this->articleUrl = $articleUrl;
        $this->articleSlug = $articleSlug;
        $this->assetsBasePath = __DIR__ . '/../public/assets';

        // Create directories if they don't exist
        $this->ensureDirectories();
    }

    /**
     * Ensures directories exist
     */
    private function ensureDirectories()
    {
        $dirs = [
            $this->assetsBasePath . '/' . $this->featuredPath,
            $this->assetsBasePath . '/' . $this->mainContentPath,
        ];

        foreach ($dirs as $dir) {
            if (!is_dir($dir)) {
                mkdir($dir, 0755, true);
                echo "✓ Creado directorio: $dir\n";
            }
        }
    }

    /**
     * Downloads an image from a URL
     */
    public function downloadImage($imageUrl, $destinationPath)
    {
        if (empty($imageUrl)) {
            return false;
        }

        // If URL is relative, convert it to absolute
        if (strpos($imageUrl, 'http') !== 0) {
            $imageUrl = $this->makeAbsoluteUrl($imageUrl);
        }

        $fullPath = $this->assetsBasePath . '/' . $destinationPath;

        // Get image content
        $imageData = @file_get_contents($imageUrl);

        if ($imageData === false) {
            echo "✗ Error downloading: $imageUrl\n";
            return false;
        }

        // Save the image
        if (file_put_contents($fullPath, $imageData) !== false) {
            echo "✓ Descargada: $destinationPath\n";
            return true;
        }

        echo "✗ Error saving: $destinationPath\n";
        return false;
    }

    /**
     * Converts a relative URL to absolute
     */
    private function makeAbsoluteUrl($relativeUrl)
    {
        $baseUrl = parse_url($this->articleUrl);
        $base = $baseUrl['scheme'] . '://' . $baseUrl['host'];

        if (strpos($relativeUrl, '/') === 0) {
            return $base . $relativeUrl;
        }

        return $base . '/' . $relativeUrl;
    }

    /**
     * Extracts images from article HTML
     */
    public function extractImagesFromHtml($html)
    {
        $images = [];

        // Find images in <img> tags
        preg_match_all('/<img[^>]+src=["\']([^"\']+)["\'][^>]*>/i', $html, $matches);

        if (!empty($matches[1])) {
            foreach ($matches[1] as $imgSrc) {
                $images[] = $imgSrc;
            }
        }

        // Find images in data-src attributes (lazy loading)
        preg_match_all('/data-src=["\']([^"\']+)["\']/i', $html, $dataMatches);

        if (!empty($dataMatches[1])) {
            foreach ($dataMatches[1] as $imgSrc) {
                if (!in_array($imgSrc, $images)) {
                    $images[] = $imgSrc;
                }
            }
        }

        return array_unique($images);
    }

    /**
     * Gets article HTML from URL
     */
    public function fetchArticleHtml()
    {
        $html = @file_get_contents($this->articleUrl);

        if ($html === false) {
            echo "✗ Error getting article HTML\n";
            return null;
        }

        return $html;
    }

    /**
     * Generates a filename based on slug and image type
     */
    public function generateImageFilename($imageUrl, $type = 'featured')
    {
        $extension = $this->getImageExtension($imageUrl);
        $filename = $this->articleSlug . '.' . $extension;

        if ($type === 'featured') {
            return $this->featuredPath . '/' . $filename;
        }

        return $this->mainContentPath . '/' . $filename;
    }

    /**
     * Gets image extension from URL
     */
    private function getImageExtension($url)
    {
        $path = parse_url($url, PHP_URL_PATH);
        $extension = pathinfo($path, PATHINFO_EXTENSION);

        // If no extension, try to detect from MIME type or use png as default
        if (empty($extension)) {
            $extension = 'png';
        }

        // Normalize common extensions
        $extension = strtolower($extension);
        if ($extension === 'jpg') {
            $extension = 'jpeg';
        }

        return $extension;
    }

    /**
     * Processes and downloads all article images
     */
    public function processArticle()
    {
        echo "=== Downloading images for: {$this->articleSlug} ===\n\n";
        echo "URL del artículo: {$this->articleUrl}\n\n";

        $html = $this->fetchArticleHtml();
        if (!$html) {
            return;
        }

        $images = $this->extractImagesFromHtml($html);

        if (empty($images)) {
            echo "No images found in article.\n";
            return;
        }

        echo "Encontradas " . count($images) . " imagen(es):\n\n";

        $downloadedImages = [];
        $imageIndex = 0;

        foreach ($images as $imageUrl) {
            echo "Processing: $imageUrl\n";

            // Determine if it's featured image (first image) or content
            $type = ($imageIndex === 0) ? 'featured' : 'main-content';
            $destinationPath = $this->generateImageFilename($imageUrl, $type);

            if ($this->downloadImage($imageUrl, $destinationPath)) {
                $downloadedImages[] = [
                    'original_url' => $imageUrl,
                    'local_path' => $destinationPath,
                    'type' => $type
                ];
                $imageIndex++;
            }

            echo "\n";
        }

        // Generate report
        $this->generateReport($downloadedImages);
    }

    /**
     * Generates a report of downloaded images
     */
    private function generateReport($downloadedImages)
    {
        $reportPath = __DIR__ . '/image-mapping-' . $this->articleSlug . '.json';

        $report = [
            'article_slug' => $this->articleSlug,
            'article_url' => $this->articleUrl,
            'downloaded_at' => date('Y-m-d H:i:s'),
            'images' => $downloadedImages
        ];

        file_put_contents($reportPath, json_encode($report, JSON_PRETTY_PRINT));
        echo "✓ Reporte guardado en: $reportPath\n";
    }
}

// Execute script
if ($argc < 3) {
    echo "Usage: php download-images.php [ARTICLE_URL] [ARTICLE_SLUG]\n";
    echo "Example: php download-images.php https://bizee.com/articles/can-a-minor-own-a-business can-a-minor-own-a-business\n";
    exit(1);
}

$articleUrl = $argv[1];
$articleSlug = $argv[2];

$downloader = new ImageDownloader($articleUrl, $articleSlug);
$downloader->processArticle();
